### PR TITLE
[google_maps_flutter] Cloud-based map styling support

### PR DESCRIPTION
**Pull request by amuramoto**
Recreates https://github.com/flutter/plugins/pull/6553 form flutter/plugins which had approvals in-progress. 

Fixes https://github.com/flutter/flutter/issues/67631

### CODE REVIEW

**Comment by amuramoto** [comment]
@stuartmorgan I believe review for this one was in progress with you in the plugins repo

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:163]: Also, once moved, according to this: https://developers.google.com/maps/documentation/cloud-customization/overview#before-you-begin it seems that cloud styling is not compatible with the old maptypestyling.

I've never used mapID before: Does having a mapID imply that the map is going to be cloud-styled? Or is cloud-styling one of several features that can be obtained with a mapID?

**Review by ditman — COMMENTED** [review]
Web bits look great, thanks for the test! I have a small nitpick, and a couple of questions. My biggest concern is the one about having separate MapIDs per platform, not sure how we want that to work. (In the current version it seems not supported directly by the plugin.)

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:291]: Alternately we could leave that to do at the application level, with some example code; having platform-based logic isn't really much different from populating a platform structure. E.g.,

```dart
String? _getMapId() {
  if (kIsWeb) {
    return someWebKey;
  } else if (Platform.isIOS) {
    return someIOSKey;
  } else if (Platform.isAndroid) {
    return someAndroidKey;
  }
  return nil;
}
```

A multi-platform config object is problematic because it prevents 3P implementations for new platforms.

**Comment by stuartmorgan** [comment]
Thanks for picking this up! I'll try to review this later this week. In the meantime, there are a ton of CI failures mostly due to Dart compilation errors that will need to be addressed. There's one from the use of both `super.key` and `: super(key: key)` on the same constructor; there may be others as well.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:291]: This approach sounds good to me! I think we can make this part of an example app or similar.

**Comment by leighajarett** [comment]
@jonasfj could you help look into the CI failures?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
I fixed the Dart compilation errors so we can get real CI runs.

File: packages/google_maps_flutter/google_maps_flutter/example/lib/map_map_id.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/map_map_id.dart]: These info texts are now removed, and example projects is forced to use iOS platform version 12:
https://github.com/flutter/packages/pull/3682/commits/a86315500e865263bad706002c9fafe94549ab8e
One problem with this could be that when flutter eventually starts using newer version (>12) of the iOS platform version by default, these example projects are still build with 12. 

Missing implementation for tests with old platform versions:
> Change the iOS implementation package to have multiple example projects so that we are testing compilation with all the SDK versions. I can look into doing this as a prequel PR; it's something I was thinking about recently in the context of another PR as well.

File: packages/google_maps_flutter/google_maps_flutter/example/android/app/src/main/AndroidManifest.xml
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/android/app/src/main/AndroidManifest.xml:2]: I don't think we want to re-add this.

File: packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart:89]: "with" -> "the"

File: packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart:90]: Blank line above this, per Effective Dart guidance about having a single-sentence first paragraph.

File: packages/google_maps_flutter/google_maps_flutter_android/example/lib/main.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/example/lib/main.dart:87]: Same notes here.

File: packages/google_maps_flutter/google_maps_flutter_ios/example/ios/Podfile
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_ios/example/ios/Podfile:1]: You'll need to merge in the latest `main` to pick up the multi-example setup I added to `google_maps_flutter_ios`.

**Comment by reidbaker** [comment]
@amuramoto is this something you still plan to work on?

**Comment by amuramoto** [comment]
> @amuramoto is this something you still plan to work on?

I believe @jokerttu and team at Codemate are handling this. Please let me know if there's anything you need from me

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart]: These are now fixed

File: packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter/example/lib/map_map_id.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/map_map_id.dart]: My understanding is that if you don't explicitly set a platform version in your Podfile, CocoaPods will use the deployment target of your project bto build the pods, otherwise it will use the platform version set in the Podfile. Minimum deployment target should affect the cocoapod versions if the pods podspec file has s.ios.deployment_target set to certain version. I bet this has been done for GoogleMaps pod.
Just tested the build with ios11, ios12 and ios13 example apps, and feature seems to work on each as it should.

File: packages/google_maps_flutter/google_maps_flutter/example/lib/map_map_id.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/map_map_id.dart:117]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapController.m
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapController.m]: Changed as requested
Also removed compile time version checks and replaced with respondWithSelector tests

File: packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/example/lib/main.dart:91]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter_android/example/lib/main.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/example/lib/main.dart]: Changed as requested

File: packages/google_maps_flutter/google_maps_flutter_ios/example/ios/Podfile
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_ios/example/ios/Podfile:1]: Merged, and iOS examples updated

File: packages/google_maps_flutter/google_maps_flutter/README.md
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/README.md:62]: Nit: Blank line above this, matching the rest of the file.

Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/README.md:62]: Nit: remove "platform"

File: packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/Podfile
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/Podfile:1]: The iOS 11 example should not be changed to use iOS 12. That's what the iOS 12 and iOS 13 examples are for.

**Review by stuartmorgan — DISMISSED** [review]

File: packages/google_maps_flutter/google_maps_flutter/README.md
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/README.md]: Fixed

File: packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/Podfile
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/Podfile]: Of course not, accidentally missed this in rebase. Change reverted.

File: packages/google_maps_flutter/google_maps_flutter/README.md
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/README.md]: Removed

**Comment by stuartmorgan** [comment]
Update from triage: this is waiting for https://github.com/flutter/packages/pull/4141 to be landed, and then the next PR split out.

**Comment by stuartmorgan** [comment]
Update from triage: The first PR has landed, and now the implementation PR can be split out.

**Comment by jokerttu** [comment]
> now the implementation PR can be split out
@stuartmorgan: https://github.com/flutter/packages/pull/4638

**Comment by stuartmorgan** [comment]
The implementation PR should submit and auto-publish shortly, which will unblock this last part.

File: packages/google_maps_flutter/google_maps_flutter/pubspec.yaml
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/pubspec.yaml:5]: 2.5.0, since it's a new feature.

**Review by stuartmorgan — APPROVED** [review]
LGTM once the version bump is adjusted to a minor version.

File: packages/google_maps_flutter/google_maps_flutter/pubspec.yaml
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter/pubspec.yaml]: Version updated to 2.5.0 https://github.com/flutter/packages/pull/3682/commits/761d06c88cbc0710742ecf9d92d78f395bb8faec

**Comment by jokerttu** [comment]
Integration tests for web platform seems to hang and timeout after and hour. This is something new; I need to check tomorrow what's causing this. Or if it is even related to this PR.

**Comment by jokerttu** [comment]
> Integration tests for web platform seems to hang and timeout after and hour. This is something new; I need to check tomorrow what's causing this. Or if it is even related to this PR.

@stuartmorgan @bparrishMines web platform tests fixed at https://github.com/flutter/packages/pull/3682/commits/fb7a3eadf4108bdfaa572a95dc3d28bdf9b78343

**Review by bparrishMines — APPROVED** [review]
LGTM

**Pull request merged**
